File: typespec/hub/subscriptions/subscriptions.ts
import {
  type BillingInterval,
  type HubPlan,
  type HubPlanOID,
  isBillingInterval,
  isHubPlan,
  requiresBillingInterval,
} from "./plans.ts";

export interface ScheduledPlanChange {
  plan_oid: HubPlanOID;
  /** Absent when the scheduled plan is `hub-free-tier`. */
  billing_interval?: BillingInterval;
}

export interface HubSubscription {
  plan_oid: HubPlanOID;
  /** Present exactly when the plan is paid. */
  billing_interval?: BillingInterval;
  current_period_start?: string;
  current_period_end?: string;
  cancel_at_period_end: boolean;
  scheduled_change?: ScheduledPlanChange;
}

export interface SetSubscriptionPlanRequest {
  plan_oid: HubPlan;
  /** Absent for `hub-free-tier`; required otherwise. Never null. */
  billing_interval?: BillingInterval;
}

/**
 * Applies the same rules as the Go companion's Validate() to an untrusted
 * value, including an explicit `billing_interval: null`. Returns the invalid
 * JSON field names.
 */
export function validateSetSubscriptionPlanRequest(value: unknown): string[] {
  if (typeof value !== "object" || value === null) {
    return ["plan_oid", "billing_interval"];
  }
  const record = value as Record<string, unknown>;
  const fields: string[] = [];
  const planOID = record.plan_oid;
  const validPlan = typeof planOID === "string" && isHubPlan(planOID);
  if (!validPlan) {
    fields.push("plan_oid");
  }
  const billingValue = record.billing_interval;
  const present = billingValue !== undefined;
  if (billingValue === null) {
    fields.push("billing_interval");
  } else if (present && !isBillingInterval(billingValue)) {
    fields.push("billing_interval");
  } else if (
    validPlan &&
    requiresBillingInterval(planOID as HubPlan) !== present
  ) {
    fields.push("billing_interval");
  }
  return fields;
}

File: typespec/hub/subscriptions/subscriptions.go
package subscriptions

import (
	"encoding/json"
	"time"
)

type ScheduledPlanChange struct {
	PlanOID PlanOID `json:"plan_oid"`

	// Absent when the scheduled plan is FreeTier.
	BillingInterval *BillingInterval `json:"billing_interval,omitempty"`
}

type HubSubscription struct {
	PlanOID PlanOID `json:"plan_oid"`

	// Present exactly when the plan is paid.
	BillingInterval    *BillingInterval     `json:"billing_interval,omitempty"`
	CurrentPeriodStart *time.Time           `json:"current_period_start,omitempty"`
	CurrentPeriodEnd   *time.Time           `json:"current_period_end,omitempty"`
	CancelAtPeriodEnd  bool                 `json:"cancel_at_period_end"`
	ScheduledChange    *ScheduledPlanChange `json:"scheduled_change,omitempty"`
}

// OptionalBillingInterval distinguishes an absent billing_interval member
// from an explicit null, which a plain *BillingInterval field cannot: both
// decode to a nil pointer, so Validate could not reject null while accepting
// absence.
type OptionalBillingInterval struct {
	Value   BillingInterval
	Present bool
	Null    bool
}

// IsZero lets the field use `json:",omitzero"`: an absent value is omitted
// from the encoded request.
func (o OptionalBillingInterval) IsZero() bool {
	return !o.Present
}

// UnmarshalJSON is called by encoding/json even for the literal null, because
// OptionalBillingInterval implements json.Unmarshaler on a non-pointer field.
func (o *OptionalBillingInterval) UnmarshalJSON(data []byte) error {
	o.Present = true
	if string(data) == "null" {
		o.Null = true
		o.Value = ""
		return nil
	}
	var value string
	if err := json.Unmarshal(data, &value); err != nil {
		return err
	}
	o.Null = false
	o.Value = BillingInterval(value)
	return nil
}

func (o OptionalBillingInterval) MarshalJSON() ([]byte, error) {
	if o.Null {
		return []byte("null"), nil
	}
	return json.Marshal(string(o.Value))
}

type SetSubscriptionPlanRequest struct {
	PlanOID         Plan                    `json:"plan_oid"`
	BillingInterval OptionalBillingInterval `json:"billing_interval,omitzero"`
}

func (r *SetSubscriptionPlanRequest) Normalize() {}

func (r SetSubscriptionPlanRequest) Validate() []string {
	fields := make([]string, 0, 2)
	validPlan := IsPlan(PlanOID(r.PlanOID))
	if !validPlan {
		fields = append(fields, "plan_oid")
	}
	switch {
	case r.BillingInterval.Null:
		fields = append(fields, "billing_interval")
	case r.BillingInterval.Present &&
		!IsBillingInterval(r.BillingInterval.Value):
		fields = append(fields, "billing_interval")
	case validPlan &&
		RequiresBillingInterval(r.PlanOID) != r.BillingInterval.Present:
		fields = append(fields, "billing_interval")
	}
	return fields
}
